// firebase-auth-provider/dist/lib/middleware.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const inversify_1 = require("inversify");
const auth_provider_1 = require("./auth_provider");
let FirebaseAuthMiddleware = class FirebaseAuthMiddleware {
    constructor(auth) {
        this.auth = auth;
        this.excludedPaths = [];
    }
    getMiddleware() {
        return this.handle.bind(this);
    }
    handle(req, res, next) {
        return __awaiter(this, void 0, void 0, function* () {
            let needsAuth = !this.excludedPaths || !(this.excludedPaths.find((path) => {
                return req.path.startsWith(path);
            }));
            if (!needsAuth) {
                next();
                return;
            }
            const token = this.readTokenFromRequest(req);
            if (!token) {
                const err = this.createError('Unauthorized, no token specified');
                return next(err);
            }
            try {
                const decodedToken = yield this.auth.verifyToken(token);
                const user = yield this.findUser(decodedToken.uid);
                if (!user) {
                    let error = this.createError('Unauthorized, no user found');
                    return next(error);
                }
                Object.assign(req, {
                    user: user
                });
            }
            catch (err) {
                console.log(err);
                let error = this.createError('Unauthorized');
                return next(error);
            }
            next();
        });
    }
    // noinspection JSMethodCanBeStatic
    readTokenFromRequest(req) {
        const header = req.headers.authorization;
        if (header && header.startsWith('Bearer')) {
            return header.split('Bearer ')[1];
        }
        if (req.query.token) {
            return req.query.token;
        }
        return null;
    }
};
FirebaseAuthMiddleware = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(auth_provider_1.FirebaseAuthProvider)),
    __metadata("design:paramtypes", [Object])
], FirebaseAuthMiddleware);
exports.FirebaseAuthMiddleware = FirebaseAuthMiddleware;
